src/enclave/enclave_oe.c

```diff
@@ -253,30 +254,55 @@ static void _read_eeid_config()
     sgxlkl_enclave_state.config = cfg;
 }
 
+#define CHECK_INSIDE(X, S)                                  \
+    {                                                       \
+        if ((X) != NULL && !oe_is_within_enclave((X), (S))) \
+            oe_abort();                                     \
+    }
+
+#define CHECK_OUTSIDE(X, S)                                  \
+    {                                                        \
+        if ((X) != NULL && !oe_is_outside_enclave((X), (S))) \
+            oe_abort();                                      \
+    }
+
 static void _copy_shared_memory(const sgxlkl_shared_memory_t* host)
 {
     const sgxlkl_enclave_config_t* cfg = sgxlkl_enclave_state.config;
 
-    /* Deep copy where necessary */
+    /* The host shared memory struct has been copied by OE (but not its
+     * contents). */
+    CHECK_INSIDE(host, sizeof(sgxlkl_shared_memory_t));
 
-    sgxlkl_shared_memory_t* enc = &sgxlkl_enclave_state.shared_memory;
-    memset(enc, 0, sizeof(sgxlkl_shared_memory_t));
+    /* Temporary, volatile copy to make sure checks aren't reordered */
+    volatile sgxlkl_shared_memory_t* enc = oe_calloc_or_die(
+        1,
+        sizeof(sgxlkl_shared_memory_t),
+        "Could not allocate memory for shared memory\n");
 
     if (cfg->io.network)
+    {
         enc->virtio_net_dev_mem = host->virtio_net_dev_mem;
+        CHECK_OUTSIDE(enc->virtio_net_dev_mem, sizeof(struct virtio_dev));
```

Comment:
`virtio_dev` structure conains many pointers to sub structures, which also contain pointers to sub structures. Are those pointers suppoed to point to buffers outside of enclave? How to check and how to make sure a malicious host can't modify the pointers after the check?

Reply:
Even a friendly host will necessarily have to change the contents - that's what the shared memory is for. We will have to add checks on every memory access to any of those pointers in the lkl submodule.

Reply:
We should check the LKL virtio drivers do this.  This is the same use as CVMs, so if they don't then we have a fairly simple attack via VirtIO on Linux on SEV and similar technologies (e.g.  Google's recent offering).  We may need to add something at the bus layer for hooks that the drivers use to check that memory is the correct kind.  All of the descriptors that we use should be SWIOTLB-owned memory.

Reply:
Relevant dicussion at https://github.com/lsds/sgx-lkl/issues/745